public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts

```diff
@@ -0,0 +1,241 @@
+  dataProvider: CompletionDataProvider
+): Promise<Completion[]> {
+  const labelNames = await dataProvider.getLabelNames(otherLabels);
+  const result: Completion[] = labelNames.map((text) => ({
+    type: 'LABEL_NAME',
+    label: text,
+    insertText: `${text}${suffix}`,
+    triggerOnInsert,
+  }));
+
+  if (addExtractedLabels) {
+    const { extractedLabelKeys } = await dataProvider.getParserAndLabelKeys(otherLabels);
+    extractedLabelKeys.forEach((key) => {
+      result.push({
+        type: 'LABEL_NAME',
+        label: `${key} (parsed)`,
+        insertText: `${key}${suffix}`,
+        triggerOnInsert,
+      });
+    });
+  }
+
+  return result;
+}
+
+async function getLabelNamesForSelectorCompletions(
+  otherLabels: Label[],
+  dataProvider: CompletionDataProvider
+): Promise<Completion[]> {
+  return getLabelNamesForCompletions('=', true, false, otherLabels, dataProvider);
+}
+
+async function getInGroupingCompletions(
+  otherLabels: Label[],
+  dataProvider: CompletionDataProvider
+): Promise<Completion[]> {
+  return getLabelNamesForCompletions('', false, true, otherLabels, dataProvider);
+}
+
+async function getAfterSelectorCompletions(
+  labels: Label[],
+  afterPipe: boolean,
+  dataProvider: CompletionDataProvider
+): Promise<Completion[]> {
+  const { extractedLabelKeys, hasJSON, hasLogfmt } = await dataProvider.getParserAndLabelKeys(labels);
+  const allParsers = new Set(['json', 'logfmt', 'pattern', 'regexp', 'unpack']);
+  const completions: Completion[] = [];
+  const prefix = afterPipe ? '' : '| ';
+  const hasLevelInExtractedLabels = extractedLabelKeys.some((key) => key === 'level');
+  if (hasJSON) {
+    allParsers.delete('json');
+    const explanation = hasLevelInExtractedLabels ? 'use to get log-levels in the histogram' : 'detected';
+    completions.push({
+      type: 'PARSER',
+      label: `json (${explanation})`,
+      insertText: `${prefix}json`,
+    });
+  }
+
+  if (hasLogfmt) {
+    allParsers.delete('logfmt');
+    const explanation = hasLevelInExtractedLabels ? 'get detected levels in the histogram' : 'detected';
+    completions.push({
+      type: 'DURATION',
+      label: `logfmt (${explanation})`,
+      insertText: `${prefix}logfmt`,
+    });
+  }
+
+  const remainingParsers = Array.from(allParsers).sort();
+  remainingParsers.forEach((parser) => {
+    completions.push({
+      type: 'PARSER',
+      label: parser,
+      insertText: `${prefix}${parser}`,
+    });
+  });
+
+  extractedLabelKeys.forEach((key) => {
+    completions.push({
+      type: 'LINE_FILTER',
+      label: `unwrap ${key} (detected)`,
+      insertText: `${prefix}unwrap ${key}`,
+    });
+  });
+
+  completions.push({
+    type: 'LINE_FILTER',
+    label: 'unwrap',
+    insertText: `${prefix}unwrap`,
+  });
+
+  completions.push({
+    type: 'LINE_FORMAT',
+    label: 'line_format',
+    insertText: `${prefix}line_format "{{.$0}}"`,
+    isSnippet: true,
+  });
+
+  return [...LINE_FILTER_COMPLETIONS, ...completions];
```

Comment:
We shouldn't offer line filters if we are after first pipe.

Reply:
🙈 this is on my poor comment again. Now, the line filters completions are not suggested after selector at all. They should be suggested after selector, but as soon as user writes first pipe `|`, the line filters completions should be not displayed anymore. You can probably revert https://github.com/grafana/grafana/pull/55391/commits/3d003ca4bcb792b440add04b883fafe637e4c4c9 and tackle this case in separate PR.

Reply:
Oh, okay. Thanks for clarifying.

Reply:
Reverted.